test/test-write.c

```diff
@@ -797,6 +797,96 @@ static void test_write_tracking() {
+    TEST_SIMPLE_WRITE("\xd3\x80\x00\x00\x00\x00\x00\x00\x00", mpack_write(&writer, (int64_t)INT64_MIN));
+
+    // uint8
+    TEST_SIMPLE_WRITE("\x00", mpack_write(&writer, (uint8_t)0));
+    TEST_SIMPLE_WRITE("\x7f", mpack_write(&writer, (uint8_t)127));
+    TEST_SIMPLE_WRITE("\xcc\xff", mpack_write(&writer, (uint8_t)UINT8_MAX));
+
+    // uint16
+    TEST_SIMPLE_WRITE("\x00", mpack_write(&writer, (uint16_t)0));
+    TEST_SIMPLE_WRITE("\x7f", mpack_write(&writer, (uint16_t)127));
+    TEST_SIMPLE_WRITE("\xcc\xff", mpack_write(&writer, (uint16_t)UINT8_MAX));
+    TEST_SIMPLE_WRITE("\xcd\xff\xff", mpack_write(&writer, (uint16_t)UINT16_MAX));
+
+    // uint32
+    TEST_SIMPLE_WRITE("\x00", mpack_write(&writer, (uint32_t)0));
+    TEST_SIMPLE_WRITE("\x7f", mpack_write(&writer, (uint32_t)127));
+    TEST_SIMPLE_WRITE("\xcc\xff", mpack_write(&writer, (uint32_t)UINT8_MAX));
+    TEST_SIMPLE_WRITE("\xcd\xff\xff", mpack_write(&writer, (uint32_t)UINT16_MAX));
+    TEST_SIMPLE_WRITE("\xce\xff\xff\xff\xff", mpack_write(&writer, (uint32_t)UINT32_MAX));
+
+    // uint64
+    TEST_SIMPLE_WRITE("\x00", mpack_write(&writer, (uint64_t)0));
+    TEST_SIMPLE_WRITE("\x7f", mpack_write(&writer, (uint64_t)127));
+    TEST_SIMPLE_WRITE("\xcc\xff", mpack_write(&writer, (uint64_t)UINT8_MAX));
+    TEST_SIMPLE_WRITE("\xcd\xff\xff", mpack_write(&writer, (uint64_t)UINT16_MAX));
+    TEST_SIMPLE_WRITE("\xce\xff\xff\xff\xff", mpack_write(&writer, (uint64_t)UINT32_MAX));
+    TEST_SIMPLE_WRITE("\xcf\xff\xff\xff\xff\xff\xff\xff\xff", mpack_write(&writer, (uint64_t)UINT64_MAX));
+
+    // float and double
+    // TODO: we just test a few floats for now. this could certainly be extended to
+    // test more values like subnormal floats, infinities, etc.
+    TEST_SIMPLE_WRITE("\xca\x00\x00\x00\x00", mpack_write(&writer, (float)0.0f));
+    TEST_SIMPLE_WRITE("\xca\x40\x2d\xf3\xb6", mpack_write(&writer, (float)2.718f));
+    TEST_SIMPLE_WRITE("\xca\xc0\x2d\xf3\xb6", mpack_write(&writer, (float)-2.718f));
+    TEST_SIMPLE_WRITE("\xcb\x00\x00\x00\x00\x00\x00\x00\x00", mpack_write(&writer, (double)0.0));
+    TEST_SIMPLE_WRITE("\xcb\x40\x09\x21\xfb\x53\xc8\xd4\xf1", mpack_write(&writer, (double)3.14159265));
+    TEST_SIMPLE_WRITE("\xcb\xc0\x09\x21\xfb\x53\xc8\xd4\xf1", mpack_write(&writer, (double)-3.14159265));
+
+    // bool
+    // TODO: when we pass direct true or false into the _Generic it seems not to emit the correct stream
+    bool b = false;
+    TEST_SIMPLE_WRITE("\xc2", mpack_write(&writer, b));
+    b = true;
+    TEST_SIMPLE_WRITE("\xc3", mpack_write(&writer, b));
+
+    // char *
+    TEST_SIMPLE_WRITE("\xc0", mpack_write(&writer, (char *)NULL));
+    TEST_SIMPLE_WRITE("\xa4""1337", mpack_write(&writer, (char *)"1337"));
+
+    // const char *
+    TEST_SIMPLE_WRITE("\xc0", mpack_write(&writer, (const char *)NULL));
+    TEST_SIMPLE_WRITE("\xa4""1337", mpack_write(&writer, (const char *)"1337"));
```

Comment:
> Would this work without casts if you add additional handlers in the generic for
> char[] and const char[]? We could test all four string variations here.

I have seen odd behaviour before when not casting, when you mpack_write(writer, "foo") clang emitted error something like: `no selection for const char [3] in expanded _Generic`